docs/certbot_authenticators.md

```diff
@@ -0,0 +1,92 @@
+
+Among the other authenticators available to certbot, the [DNS authenticators][4] are also available through this container. DNS authenticators allow to prove ownership of a domain name by serving a challenge directly through a TXT record, added in your DNS provider, which is called [DNS-01][5]. As these challenges are a stronger proof of ownership than using HTTP-01, they also allow to provision wildcard certificates, that cover a whole subdomain layer of your domain (e.g. `*.example.com` covers `<anything>.example.com`, where `<anything>` can be... well... anything that is valid for DNS).
+
+
+## Setting up the container to use DNS-01 challenges
+
+To use DNS-01 challenges, you will need to create the credentials file for the chosen authenticator.
+
+You can find information about how to configure them by following those links for the supported authenticators:
+ - [dns-cloudflare][6]
+ - [dns-cloudxns][7]
+ - [dns-digitalocean][8]
+ - [dns-dnsimple][9]
+ - [dns-dnsmadeeasy][10]
+ - [dns-gehirn][11]
+ - [dns-google][12]
+ - [dns-linode][13]
+ - [dns-luadns][14]
+ - [dns-nsone][15]
+ - [dns-ovh][16]
+ - [dns-rfc2136][17]
+ - [dns-route53][18]
+ - [dns-sakuracloud][19]
+
+You will need to setup the authenticator file at `/etc/letsencrypt/<authenticator provider>.ini`.
+E.g. for Cloudflare, we would create the file `/etc/letsencrypt/cloudflare.ini` with the following content:
+
+```ini
+# Cloudflare API token used by Certbot
+dns_cloudflare_api_token = 0123456789abcdef0123456789abcdef01234567
+```
+
+
+## Using a DNS-01 authenticator by default
+
+You can use an authenticator solving DNS-01 challenges by default by setting the `CERTBOT_AUTHENTICATOR` environment variable in your docker configuration with for value the name of the authenticator you wish to use (e.g. `dns-cloudflare`).
+
+All the certificates needing renewal or creation will then start using that authenticator. Make sure, of course, that you've setup the authenticator correctly, as described above.
+
+
+## Using a DNS-01 authenticator for specific certificates only
+
+You might want to keep using the `webroot` authenticator in most cases, but need to use a DNS-01 challenge to setup a wildcard certificate for a given domain. Or you might even have a domain setup on Route53 while your other domains are on Cloudflare, and you thus are using `dns-cloudflare` as default authenticator.
+
+In such cases, you can specify the authenticator you wish to use in the certificate path that you are setting up as `ssl_certificate_key` in your server block of the nginx configuration. In our case, if we want to use `dns-route53` for a specific certificate, we could be using the following:
+
+```
+server {
+    listen              443 ssl;
+    server_name         yourdomain.org *.yourdomain.org;
```

Comment:
Question here, do both we and Nginx support having a wildcard in the server name?

Reply:
So for the container aspect, yes we do:
```
Found mention of authenticator 'dns-cloudflare' in name 'wildcard.<REDACTED>-dns-cloudflare'
Requesting an RSA certificate for 'wildcard.<REDACTED>-dns-cloudflare' (dns-01 through dns-cloudflare)
Requesting a certificate for *.<REDACTED> and <REDACTED>

Successfully received certificate.
Certificate is saved at: /etc/letsencrypt/live/wildcard.<REDACTED>-dns-cloudflare/fullchain.pem
Key is saved at:         /etc/letsencrypt/live/wildcard.<REDACTED>-dns-cloudflare/privkey.pem
This certificate expires on 2022-02-15.
These files will be updated when the certificate renews.
NEXT STEPS:
- The certificate will need to be renewed before it expires. Certbot can automatically renew the certificate in the background, but you may need to take steps to enable that functionality. See https://certbot.org/renewal-setup for instructions.
Saving debug log to /var/log/letsencrypt/letsencrypt.log

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
If you like Certbot, please consider supporting our work by:
 * Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
 * Donating to EFF:                    https://eff.org/donate-le
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Found all the necessary files for '/etc/nginx/conf.d/30-wildcard.conf.nokey', enabling...
```

And as per nginx:
```
$ curl -vvv --connect-to ::127.0.0.1 https://test.<REDACTED>/
* Connecting to hostname: 127.0.0.1
*   Trying 127.0.0.1:443...
* Connected to 127.0.0.1 (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=*.<REDACTED>
*  start date: Nov 17 19:13:32 2021 GMT
*  expire date: Feb 15 19:13:31 2022 GMT
*  subjectAltName: host "test.<REDACTED>" matched cert's "*.<REDACTED>"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: test.<REDACTED>
> User-Agent: curl/7.74.0
> Accept: */*
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 17 Nov 2021 20:22:00 GMT
< Content-Type: text/html
< Content-Length: 10
< Last-Modified: Wed, 17 Nov 2021 20:19:58 GMT
< Connection: keep-alive
< ETag: "619563ee-a"
< Accept-Ranges: bytes
<
It works!
* Connection #0 to host 127.0.0.1 left intact
```

Reply:
Oh, I did not know that Nginx could take [wildcards in the server name](http://nginx.org/en/docs/http/server_names.html).
Yeas, you have indeed proven that this will work ;)